### ISSUE TITLE
SIGTERM not passed to jobs on `docker stop`

### ISSUE THREAD

**Issue by jonfox** [joyent/containerpilot#448]
I don't think SIGTERM is being sent to running jobs when `docker stop` is issued to the container. In a simple config with just the Consul client agent, a SIGTERM should cause the agent to leave gracefully as the `leave_on_terminate` flag defaults to true for clients, which doesn't happen. If a `kill -s SIGTERM` is issued inside the container, it shuts down gracefully as expected.

I'm using Containerpilot version 3.3.0, GitHash: e7fda21 with the following config file:

```
{
  consul: "localhost:8500",
  logging: {
    level: "{{ .LOG_LEVEL | default "INFO" }}"
  },
  jobs: [
    {
      name: "consul-agent",
      exec: ["consul", "agent",
        "-data-dir", "/consul/data",
        "-retry-join", "{{ .CONSUL }}"
        ]
    }
  ]
}
```

If I issue `docker stop`, I get the following logs:

```
2017-07-27T14:23:52.656399952Z loaded config: {"Discovery":{},"LogConfig":{"level":"DEBUG","format":"default","output":"stdout"},"StopTimeout":5,"Jobs":[{"Name":"consul-agent","Exec":["consul","agent","-data-dir","/consul/data","-retry-join","192.168.221.10"],"Port":0,"Interfaces":null,"Tags":null,"ConsulExtras":null,"Health":null,"ExecTimeout":"","Restarts":"unlimited","StopTimeout":"","When":{"Frequency":"","Source":"","Once":"","Each":"","Timeout":""}}],"Watches":null,"Telemetry":null,"Control":{"SocketPath":"/var/run/containerpilot.socket"}}
2017-07-27T14:23:52.65653791Z control: initialized router for control server
2017-07-27T14:23:52.657049987Z event: {Startup global}
2017-07-27T14:23:52.657238958Z consul-agent.Run start
2017-07-27T14:23:52.658036077Z control: serving at /var/run/containerpilot.socket
2017-07-27T14:23:52.719546861Z ==> Starting Consul agent...
2017-07-27T14:23:52.725829123Z ==> Consul agent running!
2017-07-27T14:23:52.725849376Z            Version: 'v0.9.0'
2017-07-27T14:23:52.725858647Z            Node ID: '47d0d2ba-7e26-130e-4c40-84ae12adc602'
2017-07-27T14:23:52.725862984Z          Node name: '524c516a2afb'
2017-07-27T14:23:52.725866458Z         Datacenter: 'dc1'
2017-07-27T14:23:52.72587018Z             Server: false (bootstrap: false)
2017-07-27T14:23:52.725903947Z        Client Addr: 127.0.0.1 (HTTP: 8500, HTTPS: -1, DNS: 8600)
2017-07-27T14:23:52.725908436Z       Cluster Addr: 172.17.0.2 (LAN: 8301, WAN: 8302)
2017-07-27T14:23:52.725911572Z     Gossip encrypt: false, RPC-TLS: false, TLS-Incoming: false
2017-07-27T14:23:52.72591492Z 
2017-07-27T14:23:52.725918508Z ==> Log data will now stream in as it occurs:
2017-07-27T14:23:52.725922116Z 
2017-07-27T14:23:52.725925717Z     2017/07/27 14:23:52 [INFO] serf: EventMemberJoin: 524c516a2afb 172.17.0.2
2017-07-27T14:23:52.725931767Z     2017/07/27 14:23:52 [INFO] agent: Started DNS server 127.0.0.1:8600 (tcp)
2017-07-27T14:23:52.725935072Z     2017/07/27 14:23:52 [INFO] agent: Started DNS server 127.0.0.1:8600 (udp)
2017-07-27T14:23:52.725938138Z     2017/07/27 14:23:52 [INFO] agent: Started HTTP server on 127.0.0.1:8500
2017-07-27T14:23:52.726002768Z     2017/07/27 14:23:52 [INFO] agent: Joining cluster...
2017-07-27T14:23:52.726213627Z     2017/07/27 14:23:52 [WARN] manager: No servers available
2017-07-27T14:23:52.726226413Z     2017/07/27 14:23:52 [ERR] agent: failed to sync remote state: No known Consul servers
2017-07-27T14:23:52.726373016Z     2017/07/27 14:23:52 [INFO] agent: (LAN) joining: [192.168.221.10]
2017-07-27T14:23:52.727633682Z     2017/07/27 14:23:52 [INFO] serf: EventMemberJoin: oxygen-mint 192.168.221.10
2017-07-27T14:23:52.727878873Z     2017/07/27 14:23:52 [INFO] agent: (LAN) joined: 1 Err: <nil>
2017-07-27T14:23:52.727907656Z     2017/07/27 14:23:52 [INFO] agent: Join completed. Synced with 1 initial agents
2017-07-27T14:23:52.727913561Z     2017/07/27 14:23:52 [INFO] consul: adding server oxygen-mint (Addr: tcp/192.168.221.10:8300) (DC: dc1)
2017-07-27T14:23:52.796675605Z     2017/07/27 14:23:52 [INFO] agent: Synced node info
2017-07-27T14:24:00.180634597Z event: {Shutdown global}
2017-07-27T14:24:00.180991019Z event: {Stopping consul-agent}
2017-07-27T14:24:00.181043122Z event: {Stopped consul-agent}
```

so the client is not shutdown gracefully, and indeed Consul shows a failed node. However if I issue a `kill -s SIGTERM` inside the container, I get the following logs:

```
2017-07-27T14:26:00.51569471Z loaded config: {"Discovery":{},"LogConfig":{"level":"DEBUG","format":"default","output":"stdout"},"StopTimeout":5,"Jobs":[{"Name":"consul-agent","Exec":["consul","agent","-data-dir","/consul/data","-retry-join","192.168.221.10"],"Port":0,"Interfaces":null,"Tags":null,"ConsulExtras":null,"Health":null,"ExecTimeout":"","Restarts":null,"StopTimeout":"","When":{"Frequency":"","Source":"","Once":"","Each":"","Timeout":""}}],"Watches":null,"Telemetry":null,"Control":{"SocketPath":"/var/run/containerpilot.socket"}}
2017-07-27T14:26:00.515809765Z control: initialized router for control server
2017-07-27T14:26:00.516353035Z event: {Startup global}
2017-07-27T14:26:00.516426349Z consul-agent.Run start
2017-07-27T14:26:00.517009281Z control: serving at /var/run/containerpilot.socket
2017-07-27T14:26:00.571281989Z ==> Starting Consul agent...
2017-07-27T14:26:00.577996217Z ==> Consul agent running!
2017-07-27T14:26:00.578017429Z            Version: 'v0.9.0'
2017-07-27T14:26:00.578022777Z            Node ID: 'f0f943e1-f572-10e5-a80e-6e992179c7b1'
2017-07-27T14:26:00.578040068Z          Node name: 'a7751abcb641'
2017-07-27T14:26:00.578049014Z         Datacenter: 'dc1'
2017-07-27T14:26:00.578052909Z             Server: false (bootstrap: false)
2017-07-27T14:26:00.578056669Z        Client Addr: 127.0.0.1 (HTTP: 8500, HTTPS: -1, DNS: 8600)
2017-07-27T14:26:00.578060438Z       Cluster Addr: 172.17.0.2 (LAN: 8301, WAN: 8302)
2017-07-27T14:26:00.578064273Z     Gossip encrypt: false, RPC-TLS: false, TLS-Incoming: false
2017-07-27T14:26:00.578068443Z 
2017-07-27T14:26:00.578091994Z ==> Log data will now stream in as it occurs:
2017-07-27T14:26:00.578096753Z 
2017-07-27T14:26:00.578100617Z     2017/07/27 14:26:00 [INFO] serf: EventMemberJoin: a7751abcb641 172.17.0.2
2017-07-27T14:26:00.578104628Z     2017/07/27 14:26:00 [INFO] agent: Started DNS server 127.0.0.1:8600 (udp)
2017-07-27T14:26:00.578108436Z     2017/07/27 14:26:00 [INFO] agent: Started DNS server 127.0.0.1:8600 (tcp)
2017-07-27T14:26:00.578112303Z     2017/07/27 14:26:00 [INFO] agent: Started HTTP server on 127.0.0.1:8500
2017-07-27T14:26:00.578116098Z     2017/07/27 14:26:00 [INFO] agent: Joining cluster...
2017-07-27T14:26:00.578119936Z     2017/07/27 14:26:00 [INFO] agent: (LAN) joining: [192.168.221.10]
2017-07-27T14:26:00.578130246Z     2017/07/27 14:26:00 [WARN] manager: No servers available
2017-07-27T14:26:00.578134345Z     2017/07/27 14:26:00 [ERR] agent: failed to sync remote state: No known Consul servers
2017-07-27T14:26:00.579618555Z     2017/07/27 14:26:00 [INFO] serf: EventMemberJoin: oxygen-mint 192.168.221.10
2017-07-27T14:26:00.579636959Z     2017/07/27 14:26:00 [INFO] agent: (LAN) joined: 1 Err: <nil>
2017-07-27T14:26:00.579643135Z     2017/07/27 14:26:00 [INFO] agent: Join completed. Synced with 1 initial agents
2017-07-27T14:26:00.579731005Z     2017/07/27 14:26:00 [INFO] consul: adding server oxygen-mint (Addr: tcp/192.168.221.10:8300) (DC: dc1)
2017-07-27T14:26:01.650183844Z     2017/07/27 14:26:01 [INFO] agent: Synced node info
2017-07-27T14:26:10.798886144Z     2017/07/27 14:26:10 [INFO] Caught signal:  terminated
2017-07-27T14:26:10.798948036Z     2017/07/27 14:26:10 [INFO] Gracefully shutting down agent...
2017-07-27T14:26:10.799010334Z     2017/07/27 14:26:10 [INFO] consul: client starting leave
2017-07-27T14:26:11.548419489Z     2017/07/27 14:26:11 [INFO] serf: EventMemberLeave: a7751abcb641 172.17.0.2
2017-07-27T14:26:12.179052039Z     2017/07/27 14:26:12 [INFO] Graceful exit completed
2017-07-27T14:26:12.17910627Z     2017/07/27 14:26:12 [INFO] agent: Requesting shutdown
2017-07-27T14:26:12.179121697Z     2017/07/27 14:26:12 [INFO] consul: shutting down client
2017-07-27T14:26:12.179203581Z     2017/07/27 14:26:12 [INFO] manager: shutting down
2017-07-27T14:26:12.179243798Z     2017/07/27 14:26:12 [ERR] memberlist: Failed to send gossip to 192.168.221.10:8301: write udp [::]:8301->192.168.221.10:8301: use of closed network connection
2017-07-27T14:26:12.186761828Z     2017/07/27 14:26:12 [INFO] agent: consul client down
2017-07-27T14:26:12.186802068Z     2017/07/27 14:26:12 [INFO] agent: shutdown complete
2017-07-27T14:26:12.186822742Z     2017/07/27 14:26:12 [INFO] agent: Stopping DNS server 127.0.0.1:8600 (tcp)
2017-07-27T14:26:12.186831933Z     2017/07/27 14:26:12 [INFO] agent: Stopping DNS server 127.0.0.1:8600 (udp)
2017-07-27T14:26:12.186840277Z     2017/07/27 14:26:12 [INFO] agent: Stopping HTTP server 127.0.0.1:8500
2017-07-27T14:26:12.186876584Z     2017/07/27 14:26:12 [INFO] agent: Waiting for endpoints to shut down
2017-07-27T14:26:12.187330478Z     2017/07/27 14:26:12 [INFO] agent: Endpoints down
2017-07-27T14:26:12.187422577Z     2017/07/27 14:26:12 [INFO] Exit code:  0
2017-07-27T14:26:12.190984861Z consul-agent exited without error
2017-07-27T14:26:12.191050071Z event: {ExitSuccess consul-agent}
2017-07-27T14:26:12.191092188Z consul-agent.Run end
2017-07-27T14:26:12.191135241Z consul-agent.kill
2017-07-27T14:26:12.191146574Z killing command 'consul-agent' at pid: 18
2017-07-27T14:26:12.191237724Z job exited but restart not permitted: consul-agent
2017-07-27T14:26:12.191264515Z event: {Stopping consul-agent}
2017-07-27T14:26:12.191300274Z event: {Stopped consul-agent}
```

where it is shutdown gracefully.

I have no experience with Go, but did have a look through the code anyway. I noticed in `jobs.go` there is the following function:

```
// Kill sends SIGTERM to the Job's executable, if any
func (job *Job) Kill() {
	if job.exec != nil {
		if job.exec.Cmd != nil {
			if job.exec.Cmd.Process != nil {
				job.exec.Cmd.Process.Kill()
			}
		}
	}
}
```

In contrast to the comment, it looks like `Process.Kill()` will send a SIGKILL to the process. Maybe it should be something like `Process.Signal(syscall.SIGTERM)`?

**Comment by jonfox** [comment]
Actually looking in the `app.go` Terminate function, it looks like SIGKILL is expected above & the comment is wrong, so maybe there should be a job Terminate function to send SIGTERM which gets called immediately in app Terminate, and a default StopTimeout > 0?

**Comment by tgross** [comment]
Some of the understandable confusion over the code is that there's a difference between killing a _job_ and killing the job's _exec_. Killing the job means to kill it's `exec` process but also to stop trying to restart it, stop tracking its state, deregister it from Consul, etc. Whereas killing the job's `exec` is just the forked process.

When we call `App.Terminate()`, the jobs all get the shutdown signal and are responsible for cleaning themselves up, but then we turn around and SIGKILL all the processes intentionally. So what we need to figure out is why the job isn't cleaning up the process gracefully.

It's a little bit of a circuitous path to get there thru the code base because of the use of `Context`, but the way the individual job `exec` get shutdown is that when the job receives the appropriate event (like a global shutdown), we run the `cancelFunc` of the `Context` during `Job.cleanup` ([here](https://github.com/joyent/containerpilot/blob/818edfb2306c30df7b4f7062386a7d84e4fcee5c/jobs/jobs.go#L344)). By canceling the `Context` we pass into the `commands.go:Command`, we stop blocking on `<- ctx.Done` [here](https://github.com/joyent/containerpilot/blob/818edfb2306c30df7b4f7062386a7d84e4fcee5c/commands/commands.go#L70), and that calls `Command.Kill()`, which sends a SIGKILL to the process group [here](https://github.com/joyent/containerpilot/blob/818edfb2306c30df7b4f7062386a7d84e4fcee5c/commands/commands.go#L114)

So long story short is that I agree that this is a bug in `Command.Kill()`. At a quick glance it looks like we should be able simply swap this line out for `syscall.Kill(-c.Cmd.Process.Pid, syscall.SIGTERM)`. I'll open a PR for that shortly.

**Comment by tgross** [comment]
> At a quick glance it looks like we should be able simply swap this line out for syscall.Kill(-c.Cmd.Process.Pid, syscall.SIGTERM). I'll open a PR for that shortly.

Ok, well the problem with doing that is that it only covers the case of graceful shutdown. It breaks cases where we have to force a shutdown because of a timeout in the job's `exec` (or health check `exec`). So maybe rather than replacing `Command.Kill()` we want to inject a graceful shutdown into `Job.cleanup`.

**Comment by tgross** [comment]
Opened https://github.com/joyent/containerpilot/pull/449

**Comment by jonfox** [comment]
Thanks for getting onto this so quickly, I look forward to testing the release.

Also thanks for the code explanation. It's clearly time I got round to learning Go!

**Comment by tgross** [comment]
Fix released in https://github.com/joyent/containerpilot/releases/tag/3.3.1
